util/dvsim/FpvCfg.py

```diff
@@ -0,0 +1,277 @@
+        else:
+            cov_table.append([
+                fpv_coverage["stimuli"],
+                fpv_coverage["coi"],
+                fpv_coverage["proof"]
+            ])
+            summary.append(fpv_coverage["stimuli"])
+            summary.append(fpv_coverage["coi"])
+            summary.append(fpv_coverage["proof"])
+
+            if len(cov_table) > 1:
+                results_str = tabulate(cov_table, headers="firstrow",
+                                       tablefmt="pipe", colalign=cov_colalign)
+
+            else:
+                results_str = "No content in fpv_coverage\n"
+                summary = ["N/A", "N/A", "N/A"]
+        return results_str, summary
+
+    def gen_results_summary(self):
+        # Gathers the aggregated results from all sub configs
+        # The results_summary will only contain the passing rate and
+        # percentages of the stimuli, coi, and proven coverage
+        # The email_summary will contain all the information from results_md
+        log.info("Create summary of FPV results")
+
+        results_str = "## " + self.results_title + " (Summary)\n\n"
+        results_str += "### " + self.timestamp_long + "\n\n"
+
+        colalign = ("center", ) * len(self.summary_header)
+        table = [self.summary_header]
+        for cfg in self.cfgs:
+            try:
+                table.append(cfg.result_summary[cfg.name])
+            except KeyError as e:
+                table.append([cfg.name, "ERROR", "N/A", "N/A", "N/A"])
+                log.error("cfg: %s could not find generated results_summary: %s", cfg.name, e)
+        if len(table) > 1:
+            self.results_summary_md = results_str + tabulate(
+                table, headers="firstrow", tablefmt="pipe", colalign=colalign)
+        else:
+            self.results_summary_md = results_str
+
+        log.info("[result summary]: %s", self.results_summary_md)
+
+        # Generate email results summary
+        colalign = ("left", ) + ("center", ) * (len(self.header) - 1)
+        email_table = [self.header]
+        error_message = ""
+
+        for cfg in self.cfgs:
+            email_result = cfg.result.get("fpv_summary")
+            if email_result is not None:
+                email_table.append([
+                    cfg.name,
+                    str(email_result["fpv_errors"]) + " E ",
+                    str(email_result["fpv_warnings"]) + " W ",
+                    str(email_result["proven"]) + " G ",
+                    str(email_result["cex"]) + " E ",
+                    str(email_result["undetermined"]) + " W ",
+                    str(email_result["covered"]) + " G ",
+                    str(email_result["unreachable"]) + " E ",
+                    email_result["pass_rate"],
+                    email_result["cov_rate"]
+                ])
+            messages = cfg.result.get("messages")
+            if messages is not None:
+                # TODO: clean up FPV warnings, then enable warning disable
```

Comment:
Left-over TODO message. I don't think I quite understand it: are you planning to follow up in a later PR?

Reply:
Yeah, currently I am trying to not display FPV warning messages because some DUTs might have hundreds of them. I would like to clean up the warnings in a separate PR first, then remove the extra ["fpv_warnings"] input from line 169